test/expr/unary/keypath/keypath.swift

```diff
@@ -1067,8 +1068,10 @@ func f_56996() {
 // https://github.com/apple/swift/issues/55805
 // Key-path missing optional crashes compiler: Inactive constraints left over?
 func f_55805() {
-  let _: KeyPath<String?, Int?> = \.utf8.count // expected-error {{value of optional type 'String.UTF8View?' must be unwrapped to refer to member 'count' of wrapped base type 'String.UTF8View'}}
-  // expected-note@-1 {{chain the optional using '?' to access member 'count' only for non-'nil' base values}}
+  let _: KeyPath<String?, Int?> = \.utf8.count // expected-error {{key path value type 'Int' cannot be converted to contextual type 'Int?'}}
+  // expected-error@-1 {{key path root inferred as optional type 'String?' must be unwrapped to refer to member 'utf8' of unwrapped type 'String'}}
+  // expected-note@-2 {{chain the optional using '?.' to access unwrapped type member 'utf8'}}
+  // expected-note@-3 {{unwrap the optional using '!.' to access unwrapped type member 'utf8'}}
```

Comment:
So for this given example we should produce only one diagnostic related to missing unwrap of `a`:

```
struct X {
  var a: A?
}

struct A {
  var v: Int
}

let _: KeyPath<X, Int?> = \.a?.v
```

But I think with this change we'd produce two errors? 

I think the problem is that "root" unwrapping is different from member unwrapping i.e. we cannot spell something like `\Optional<[Int]>?.count` or `\?.utf8.count`.

So maybe change should actually be limited to member references where its base type is key path root...